src/packet_sources/macos.rs

```diff
@@ -0,0 +1,223 @@
+    buf: [u8; IPC_BUF_SIZE],
+    net_tx: Sender<NetworkEvent>,
+    net_rx: Receiver<NetworkCommand>,
+    conf_rx: UnboundedReceiver<ipc::FromProxy>,
+    sd_watcher: broadcast::Receiver<()>,
+}
+
+#[async_trait]
+impl PacketSourceTask for MacosTask {
+    async fn run(mut self) -> Result<()> {
+        log::debug!("Waiting for IPC connection...");
+        log::debug!("IPC connected!");
+
+        loop {
+            tokio::select! {
+                // wait for graceful shutdown
+                _ = self.sd_watcher.recv() => break,
+                // pipe through changes to the intercept list
+                Some(cmd) = self.conf_rx.recv() => {
+                    assert!(matches!(cmd, ipc::FromProxy { message: Some(ipc::from_proxy::Message::InterceptSpec(_)) }));
+                    cmd.encode(&mut self.buf.as_mut_slice())?;
+                    self.ipc_server.filter_tx.try_write(&self.buf)?;
+                    println!("SetIntercept {:?}", cmd);
+                },
+                // read packets from the IPC pipe into our network stack.
+                _ = self.ipc_server.ip_rx.readable() => {
+                    match self.ipc_server.ip_rx.try_read(&mut self.buf){
+                        Ok(len) => {
+                            if len == 0 {
+                                return Err(anyhow!("redirect daemon exited prematurely."));
+                            }
+                            let mut cursor = Cursor::new(&self.buf[..len]);
+                            let Ok(FromRedirector { message: Some(message)}) = FromRedirector::decode(&mut cursor) else {
+                                return Err(anyhow!("Received invalid IPC message: {:?}", &self.buf[..len]));
+                            };
+                            assert_eq!(cursor.position(), len as u64);
+                            let (data, process_name) = match message {
+                                Packet(PacketWithMeta { data, process_name, ..}) => (data, process_name.map(PathBuf::from)),
+                            };
+                            let Ok(mut packet) = IpPacket::try_from(data) else {
+                                log::error!("Skipping invalid packet: {:?}", &self.buf[..len]);
+                                continue;
+                            };
+                            packet.fill_ip_checksum();
+                            let event = NetworkEvent::ReceivePacket {
+                                packet,
+                                tunnel_info: TunnelInfo::Macos {
+                                    process_name,
+                                },
+                            };
+                            if self.net_tx.try_send(event).is_err() {
+                                log::warn!("Dropping incoming packet, TCP channel is full.")
+                            };
+                        },
+                        Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => continue,
```

Comment:
yes because it could be a false positive [[docs](https://docs.rs/tokio/latest/tokio/net/unix/pipe/struct.Receiver.html#:~:text=The%20function%20may,are%20not%20set.)]